packages/selection/src/useSelection.ts

```diff
@@ -148,42 +148,13 @@ function stateReducer(state: IUseSelectionState<any>, action: SELECTION_ACTION<a
       return { ...state, focusedItem: action.items[action.items.length - 1] };
     }
     case 'MOUSE_SELECT': {
-      let isSelectControlled = false;
-      let isFocusControlled = false;
-
-      if (action.onSelect) {
-        action.onSelect(action.payload);
-        isSelectControlled = true;
-      }
-
-      if (action.onFocus) {
-        action.onFocus(undefined);
-        isFocusControlled = true;
-      }
-
-      if (isFocusControlled && isSelectControlled) {
-        return state;
-      }
-
-      const updatedState = { ...state };
-
-      if (!isSelectControlled) {
-        updatedState.selectedItem = action.payload;
-      }
-
-      if (!isFocusControlled) {
-        updatedState.focusedItem = undefined;
-      }
-
-      return updatedState;
+      return {
+        ...state,
+        selectedItem: action.payload,
+        focusedItem: undefined
+      };
     }
     case 'KEYBOARD_SELECT': {
-      if (action.onSelect) {
-        action.onSelect(action.payload);
-
-        return state;
```

Comment:
Does this work for both controlled and uncontrolled usages? This removes the [ability for us to bail out of the dispatch action](https://reactjs.org/docs/hooks-reference.html#bailing-out-of-a-dispatch).

Reply:
I think it should work for both controlled and uncontrolled usages. Only uncontrolled usages use the internal state reducer, so I think technically this is what we want:

```jsx
onSelect && onSelect(item);
onFocus && onFocus();

if (isControlled === false) { // or some variant of this logic
  dispatch({ type: 'MOUSE_SELECT', payload: item });
}
```

What do you think?

Reply:
👍 As long it avoids the double-render than we should be good

Reply:
Regarding my comment above: That would involve invasive changes that could be reserved for a separate PR. 

Currently, controlled vs uncontrolled is determined by a check on `onSelect` or `onFocus`. This is incorrect because [controlled vs uncontrolled](https://gist.github.com/ryanflorence/e2fa045ad523f2228d34ce3f94df75b3) should be determined by a controlled prop value. [This snippet](https://github.com/reach/reach-ui/blob/develop/packages/utils/src/index.tsx#L456) shows how Ryan Florence does it:

```jsx
const isControlled = controlledProp != null
```

Uncontrolled usages are driven by state, while controlled usages are driven by props. The `useSelection` hook will need to be updated to reflect that. But that is a separate issue from https://github.com/zendeskgarden/react-components/issues/891.

Reply:
When we update the controlled variables, lets use the following logic to align with native elements. Null is a valid controlled value in some use-cases.

```
const isControlled = controlledProp !== undefined;
```

Reply:
> Null is a valid controlled value in some use-cases

What do you think about using a strict equality [check against both](https://github.com/zendeskgarden/react-containers/blob/main/packages/accordion/src/useAccordion.ts#L62) `null` and `undefined`? A `null` value should indicate uncontrolled usage, and if it doesn't there's likely an issue with that component.

In React, a prop that receives `undefined` or `null` is considered an uncontrolled prop:

```jsx
// React considers this a controlled component
<input type="text" value="Hello world" /> 

// React considers this a uncontrolled component
<input type="text" value={null} />

// React considers this a uncontrolled component
<input type="text" value={undefined} />
```

Reply:
I think that's the web determining that, not React. I don't think we should count `null` as uncontrolled.

Uncontrolled means a consumer hasn't provided a value, i.e. not defined. Null [represents the intentional absence of any object value](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/null).

Downshift has an [isControlledProp](https://github.com/downshift-js/downshift/blob/f4cb1779fdca7973276cd2b5f217203e3b1c0c25/src/utils.js#L270-L272) utility which includes the logic that I've seen most often (and what we use in `react-components` currently).

Reply:
> I think that's the web determining that, not React. 

How is the web determining that? Would I be able to determine this in a vanilla `.html` file?

> Downshift has an isControlledProp utility which includes the logic that I've seen most often (and what we use in react-components currently).

I think that's an oversight on Downshift's part. 

If open-source libraries are used as evidence, then I would point to Reach UI's [isControlled](https://github.com/reach/reach-ui/blob/develop/packages/utils/src/index.tsx#L456). But for me, the more reliable evidence is found in React source code for its input implementation, [isControlled](https://github.com/facebook/react/blob/c601f7a64640290af85c9f0e33c78480656b46bc/packages/react-dom/src/client/ReactDOMFiberInput.js#L39).

Reply:
For native inputs and Reach UI I believe they only accept strings as their controlled values. For that use-case I could see null/undefined meaning the same thing because an empty value can be provided with `''`

With us allowing objects for controlled values I think null has a distinct meaning for us in comparison. 

This change would also be widespread across all hooks and Garden elements and would constitute a breaking change since certain controlled instances could break for consumers. I still agree that the original logic of basing controlled state from the `onChange` handlers should be changed.

Reply:
I agree that this PR should not introduce breaking changes. I think those are some valid considerations and would like to discuss controlled vs uncontrolled more after this PR.